app/frontend/src/features/userQueries/useUsers.ts

```diff
@@ -1,21 +1,32 @@
 import { Error } from "grpc-web";
-import { useEffect } from "react";
+import { useCallback, useEffect, useRef } from "react";
 import { useQueries, useQueryClient } from "react-query";
 import { User } from "../../pb/api_pb";
 import { service } from "../../service";
+import { arrayEq } from "../../utils/arrayEq";
 import { userStaleTime } from "./constants";
 
 export default function useUsers(ids: number[], invalidate: boolean = false) {
   const queryClient = useQueryClient();
-  useEffect(() => {
+  const idsRef = useRef(ids);
+  const handleInvalidation = useCallback(() => {
     if (invalidate) {
       queryClient.invalidateQueries({
         predicate: (query) =>
           query.queryKey[0] === "user" &&
-          ids.includes(query.queryKey[1] as number),
+          idsRef.current.includes(query.queryKey[1] as number),
       });
     }
-  }, [invalidate, queryClient, ids]);
+  }, [invalidate, queryClient]);
+  useEffect(() => {
+    handleInvalidation();
+  }, [handleInvalidation]);
+
+  //arrays use reference equality, so you can't use ids in useEffect directly
+  if (!arrayEq(idsRef.current, ids)) {
+    idsRef.current = ids;
+    handleInvalidation();
+  }
```

Comment:
What do you think of this solution? Although I doubt a component will ever explicitly request a certain array of ids, I felt weird putting this kind of 'requirement' in which wouldn't cause any errors other than endless render loops.

Reply:
How about putting it in another `useEffect` with no dependency array?

Reply:
The `handleInvalidation` in the if-statement? That breaks the rule of hooks since you can't use one conditionally.

After more thoughts, I think this might be an ok compromise if we make `handleInvalidation` an async function? If it works with `useEffect`, then it should be ok?

While what I said works, the additional thoughts made me think that it's still a bit awkward/unintuitive to have to remember in all consumers to pass in a memoised/stable-ish list into the hook rather than using it literally as I'd naturally expect.

Reply:
I mean the whole if statement. I think that's the way to go.